pkgs/build-support/setup-hooks/auto-patchelf.py

```diff
@@ -223,24 +246,24 @@ def auto_patchelf_file(path: Path, runtime_deps: list[Path], append_rpaths: List
     # failing at the first one, because it's more useful when working
     # on a new package where you don't yet know the dependencies.
     for dep in file_dependencies:
-        if dep.is_absolute() and dep.is_file():
-            # This is an absolute path. If it exists, just use it.
-            # Otherwise, we probably want this to produce an error when
-            # checked (because just updating the rpath won't satisfy
-            # it).
-            continue
-        elif (libc_lib / dep).is_file():
-            # This library exists in libc, and will be correctly
-            # resolved by the linker.
-            continue
+        found_dependency = None
+        for candidate in dep:
+            if candidate.is_absolute() and candidate.is_file():
+                # This is an absolute path. If it exists, just use it.
+                # Otherwise, we probably want this to produce an error when
+                # checked (because just updating the rpath won't satisfy it).
+                continue
 
-        if found_dependency := find_dependency(dep.name, file_arch, file_osabi):
-            rpath.append(found_dependency)
-            dependencies.append(Dependency(path, dep, True))
-            print(f"    {dep} -> found: {found_dependency}")
-        else:
-            dependencies.append(Dependency(path, dep, False))
-            print(f"    {dep} -> not found!")
+            if found_dependency := find_dependency(candidate.name, file_arch, file_osabi):
+                rpath.append(found_dependency)
+                dependencies.append(Dependency(path, candidate, found=True))
+                print(f"    {candidate} -> found: {found_dependency}")
+                break
+
+        if found_dependency is None:
+            dep_name = dep[0] if len(dep) == 1 else f"any({', '.join(map(str, dep))})"
+            dependencies.append(Dependency(path, dep_name, found=False))
+            print(f"    {dep_name} -> not found!")
```

Comment:
While I do understand the need to move to lists of strings (as the FreeDesktop spec uses lists), this rewrite seems incomplete and bogus to me.

1. The logic around libc_lib has been completely removed. Please re-add it.
2. At this stage, I think it is necessary to introduce two data types for dependencies. Or to split the logic looking for these dependencies. While libc_lib only applies to DT_NEEDED, absolute paths in .note.dlopen are probably not a `continue` but a break.
3. absolute paths should be considered as found. The continue statement doe snot work inside the nested loop over candidates.

Splitting this into subfunctions or methods of the class of each type of dependency may help.
I am not too much into style ultimately. I care more about the logic and it seems this code broke it.

Reply:
@layus So, the reason for the removal of the glibc thing was because `glibc` was being removed from the rpath of `systemd` binaries altogether, even when they had `DT_NEEDED` dependencies on it. This resulted in e.g. `/nix/store/yggl0850psl4anbavqqfg5hgpdsc2196-extra-utils/bin/udevadm: error while loading shared libraries: libacl.so.1: cannot open shared object file: No such file or directory`

It seemed odd to me, and I wasn't sure why `glibc` was being excluded in the first place. Running builds for my systems on hydra, however, I am now seeing at least a symptom of why:

```
auto-patchelf: 9 dependencies could not be satisfied
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/elf/testdata/gcc-amd64-linux-exec
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/line-gcc-zstd.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/line-gcc-dwarf5.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/ranges.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/typedef.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/line-clang-dwarf5.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/line-gcc.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/line-clang.elf
error: auto-patchelf could not satisfy dependency libc.so.6 wanted by /nix/store/rb3vfzranc26i7fk24ikm6v6lzg7v9cz-go-1.21.0-linux-amd64-bootstrap/share/go/src/debug/dwarf/testdata/split.elf
auto-patchelf failed to find all the required dependencies.
Add the missing dependencies to --libs or use `--ignore-missing="foo.so.1 bar.so etc.so"`.
```

The odd bit is that this seems only relevant to the debug files; which is extra odd because systemd uses `separateDebugInfo` and the debug files for it don't have this problem.

Can you explain the logic for this `glibc` exclusion? It's very confusing to me.  And the consequence of failing to find `libacl.so.1` (and probably others) for systemd binaries is rather critical.